### ISSUE TITLE
add: --to-remote to gs fails when not using a service account to authenticate

### ISSUE THREAD

**Issue by KeyboardSounds** [iterative/dvc#5509]
# Bug Report

## Description

`dvc add --to-remote` fails when the remote is a gs bucket, and the user is using credentials from `gcloud auth application-default login` rather than a service account. Here's the error message:
```
google.auth.exceptions.TransportError: Failed to retrieve http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/?recursive=true from the Google Compute Enginemetadata service. Compute Engine Metadata server unavailable
```

This function is where the error happens:

https://github.com/iterative/dvc/blob/d587d41d6c73147be19fe207c9d0881392417f8f/dvc/fs/gs.py#L80-L105

From poking around in Google's [python-storage](https://github.com/googleapis/python-storage/blob/master/google/cloud/storage/bucket.py), [google-auth-library-python](https://github.com/googleapis/google-auth-library-python/blob/6407258956ec42e3b722418cb7f366e5ae9272ec/google/auth/compute_engine/credentials.py) and [google-auth-library-python](https://github.com/googleapis/google-auth-library-python/blob/master/google/auth/compute_engine/_metadata.py) (those links are to some of the relevant code, not just the repo home page), I suspect the issue is that for a signed url to be generated, the signing credentials must be those of a service account.

There might not be a solution to this, but at least users could be warned about this issue in the docs. The error message could potentially be clearer as well. The url referred to in the error message: `http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/?recursive=true` would only be accessible from a GCE instance, I believe.

### Reproduce

1. `gcloud auth application-default login`. Make sure `GOOGLE_APPLICATION_CREDENTIALS` is not set.
2. `dvc add --to-remote --file dataset.dvc ../external-dataset.zip --verbose` where the default remote is a gs bucket.

### Expected

The `dvc add` command succeeds without errors.

### Environment information

**Output of `dvc version`:**

```console
$ dvc version
DVC version: 2.0.0a0+d587d4
---------------------------------
Platform: Python 3.7.3 on Darwin-20.3.0-x86_64-i386-64bit
Supports: All remotes
Cache types: <https://error.dvc.org/no-dvc-cache>
Caches: local
Remotes: gs
Workspace directory: apfs on /dev/disk1s5s1
Repo: dvc, git
```

**Additional Information (if any):**

Verbose output:
```
▶ dvc add --to-remote --file dataset.dvc ../external-dataset.zip --verbose
2021-02-23 14:58:27,565 DEBUG: Check for update is enabled.
Adding...
2021-02-23 15:06:03,749 ERROR: unexpected error - Failed to retrieve http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/?recursive=true from the Google Compute Enginemetadata service. Compute Engine Metadata server unavailable
------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/main.py", line 50, in main
    ret = cmd.run()
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/command/add.py", line 32, in run
    jobs=self.args.jobs,
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/repo/__init__.py", line 49, in wrapper
    return f(repo, *args, **kwargs)
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/repo/scm_context.py", line 14, in run
    return method(repo, *args, **kw)
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/repo/add.py", line 131, in add
    **kwargs,
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/repo/add.py", line 172, in _process_stages
    command="add",
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/data_cloud.py", line 111, in transfer
    return remote.transfer(from_fs, from_fs.path_info, jobs=jobs)
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/remote/base.py", line 520, in transfer
    no_progress_bar=no_progress_bar,
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/objects/__init__.py", line 313, in transfer
    odb.add(tmp_info, odb.fs, hash_info)
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/objects/db/base.py", line 47, in add
    self.check(hash_info)
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/objects/db/base.py", line 110, in check
    obj.check(self)
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/objects/__init__.py", line 39, in check
    actual = get_hash(self.path_info, self.fs, odb.fs.PARAM_CHECKSUM)
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/objects/stage.py", line 137, in get_hash
    hash_info = get_file_hash(path_info, fs, name)
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/objects/stage.py", line 24, in get_file_hash
    name, file_md5(path_info, fs), size=fs.getsize(path_info)
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/utils/__init__.py", line 50, in file_md5
    binary = not istextfile(fname, fs=fs)
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/istextfile.py", line 31, in istextfile
    with fs.open(fname, "rb") as fobj:
  File "/usr/local/anaconda3/lib/python3.7/contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/utils/http.py", line 16, in open_url
    with iter_url(url, **iter_opts) as (response, it):
  File "/usr/local/anaconda3/lib/python3.7/contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/utils/http.py", line 59, in iter_url
    response = request()
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/utils/http.py", line 32, in request
    the_url = url() if callable(url) else url
  File "/usr/local/anaconda3/lib/python3.7/site-packages/dvc/fs/gs.py", line 103, in _generate_download_url
    auth_request, ""
  File "/usr/local/anaconda3/lib/python3.7/site-packages/google/auth/compute_engine/credentials.py", line 230, in __init__
    sa_info = _metadata.get_service_account_info(request)
  File "/usr/local/anaconda3/lib/python3.7/site-packages/google/auth/compute_engine/_metadata.py", line 234, in get_service_account_info
    return get(request, path, params={"recursive": "true"})
  File "/usr/local/anaconda3/lib/python3.7/site-packages/google/auth/compute_engine/_metadata.py", line 165, in get
    "metadata service. Compute Engine Metadata server unavailable".format(url)
google.auth.exceptions.TransportError: Failed to retrieve http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/?recursive=true from the Google Compute Enginemetadata service. Compute Engine Metadata server unavailable
------------------------------------------------------------
2021-02-23 15:06:04,985 DEBUG: Version info for developers:
DVC version: 2.0.0a0+d587d4
---------------------------------
Platform: Python 3.7.3 on Darwin-20.3.0-x86_64-i386-64bit
Supports: All remotes
Cache types: <https://error.dvc.org/no-dvc-cache>
Caches: local
Remotes: gs
Workspace directory: apfs on /dev/disk1s5s1
Repo: dvc, git
```

**Comment by efiop** [comment]
Hi @KeyboardSounds !

Thanks for the feedback! Likely going to be fixed by https://github.com/iterative/dvc/issues/5500

**Comment by isidentical** [comment]
I can verify that #5500 resolves

**Comment by isidentical** [comment]
Resolved by #5526.
